### PR TITLE
Draft PR: adding support for pure cert/key authentication ('no proxy' mode)

### PR DESCRIPTION
**Pull request by gganis**
Client enforces this mode by setting XrdSecGISNOPROXY = 1.
Key file must be pass-less for this work.
NB: if the key-file is pass-less and XrdSecGISNOPROXY = 0 (default) authentication still works with the usual protocol, i.e. creating a proxy and using that for the handshake. Setting XrdSecGISNOPROXY = 1 avoids those additional steps.

### CODE REVIEW

**Comment by bbockelm** [comment]
Hi @gganis!

@abh3 asked me to take a quick look at this.  I don't have the time right now to do a thorough test so first impressions from reading the code:

0.  The double-negative (disabling no proxy) is a bit confusing.  Maybe default to `XrdGSICREATEPROXY=1` and ask folks to turn it off?
1.  What's the latest coding style guides for xrootd?  In particular, I see a few places where `0` is used instead of `nullptr`.  They're effectively equivalent in this use case but I find `nullptr` more explicit / modern styling.
2. It seems that "proxy" versus "cert/key" modes are mutually exclusive.  That is, if "no proxy" mode is activated then a cert/key is _required_ (and otherwise cert/key are never accepted.

Item (2) seems relatively restrictive.  Other clients, such as the traditional `globus-*` ones, will cleanly fall back to the cert/key if a proxy isn't present, no?

I don't know the driving use case here but it seems this would be more familiar:
1.  If `XrdSecGSICREATEPROXY=1` (default), a proxy is auto-generated from the cert/key pair if one is not found.
2. If `XrdSecGSICREATEPROXY=0`, a proxy is used if present.  Otherwise, the cert/key pair is used if present.

Finally - what are the forward / backward compatibility concerns here?  Can older servers handle a cert directly?

**Comment by simonmichal** [comment]
Hi @gganis, thanks for the PR, I noticed that whether the XrdSecGISNOPROXY is set or not the proxy certificate is created, is this intended?

1.
```
$ export XrdSecGISNOPROXY=1
$ voms-proxy-info

Couldn't find a valid proxy.

$ xrdfs hostname ls /tmp
210830 14:35:45 9668 cryptossl_X509CreateProxy: Your identity: /DC=ch/DC=cern/OU=Organic Units/OU=Users/CN=simonm/CN=682103/CN=Michal Kamil Simon
/tmp/.ICE-unix
/tmp/.Test-unix
/tmp/.X11-unix
/tmp/.XIM-unix
/tmp/.font-unix
/tmp/Makefile
/tmp/systemd-private-7cfdb59adbf447adb788956d19da928c-cups.service-tB8Gvt
/tmp/systemd-private-7cfdb59adbf447adb788956d19da928c-ntpd.service-Yc9Pxr
/tmp/xrootd
$ voms-proxy-info
subject   : /DC=ch/DC=cern/OU=Organic Units/OU=Users/CN=simonm/CN=682103/CN=Michal Kamil Simon/CN=234434324
issuer    : /DC=ch/DC=cern/OU=Organic Units/OU=Users/CN=simonm/CN=682103/CN=Michal Kamil Simon
identity  : /DC=ch/DC=cern/OU=Organic Units/OU=Users/CN=simonm/CN=682103/CN=Michal Kamil Simon
type      : RFC compliant proxy
strength  : 512 bits
path      : /tmp/x509up_u1000
timeleft  : 11:59:57
```

2.
```
$ export XrdSecGISNOPROXY=0
$ voms-proxy-info

Couldn't find a valid proxy.
$ ./xrdfs slc7-test ls /tmp
210830 14:46:20 10621 cryptossl_X509CreateProxy: Your identity: /DC=ch/DC=cern/OU=Organic Units/OU=Users/CN=simonm/CN=682103/CN=Michal Kamil Simon
/tmp/.ICE-unix
/tmp/.Test-unix
/tmp/.X11-unix
/tmp/.XIM-unix
/tmp/.font-unix
/tmp/Makefile
/tmp/systemd-private-7cfdb59adbf447adb788956d19da928c-cups.service-tB8Gvt
/tmp/systemd-private-7cfdb59adbf447adb788956d19da928c-ntpd.service-Yc9Pxr
/tmp/xrootd
$ voms-proxy-info
subject   : /DC=ch/DC=cern/OU=Organic Units/OU=Users/CN=simonm/CN=682103/CN=Michal Kamil Simon/CN=1763077269
issuer    : /DC=ch/DC=cern/OU=Organic Units/OU=Users/CN=simonm/CN=682103/CN=Michal Kamil Simon
identity  : /DC=ch/DC=cern/OU=Organic Units/OU=Users/CN=simonm/CN=682103/CN=Michal Kamil Simon
type      : RFC compliant proxy
strength  : 512 bits
path      : /tmp/x509up_u1000
timeleft  : 11:59:58
```

File: src/XrdSecgsi/XrdSecProtocolgsi.cc [hunk 1]
```diff
@@ -5082,12 +5111,22 @@ int XrdSecProtocolgsi::QueryProxy(bool checkcache, XrdSutCache *cache,
...
-            int nci = (*ParseFile)(pi->out, po->chain);
-            if (nci < 2) {
-               DEBUG("proxy files must have at least two certificates"
-                     " (found: "<<nci<<")");
-               continue;
+            if (pi->nopxy) {
+               // Parse the cert file
+               int nci = (*ParseFile)(pi->cert, po->chain, pi->key);
+               if (nci < 1) {
+                  DEBUG("proxy files must have at least 1 certificates"
```
Comment (simonmichal) [inline src/XrdSecgsi/XrdSecProtocolgsi.cc:5118]: Is this log message correct, aren't we parsing here the certificate (not the proxy)?

**Comment by simonmichal** [comment]
I looks OK in valgrind.

**Comment by gganis** [comment]
Hi @bbockelm, @simonmichal,
Sorry for the late reply, I am just back from vacation.
Thank you for you comments/questions. I will go through and come back to you asap.

File: src/XrdSecgsi/XrdSecProtocolgsi.cc (hunk 1 above)
Comment (gganis) [reply src/XrdSecgsi/XrdSecProtocolgsi.cc]: Correct. Fixed. Thanks!

**Comment by gganis** [comment]
The logic has been reversed following Brian's suggestion.
Backward compatibility checks with version 5.3.1, 4.12.8 are ok.

**Comment by abh3** [comment]
Superseded by pull request 1561.

**Pull request closed without merging**